presto-hive/src/main/java/com/facebook/presto/hive/HiveTableProperties.java

```diff
@@ -240,22 +241,65 @@ public static HiveStorageFormat getHiveStorageFormat(Map<String, Object> tablePr
         List<String> bucketedBy = getBucketedBy(tableProperties);
         List<SortingColumn> sortedBy = getSortedBy(tableProperties);
         int bucketCount = (Integer) tableProperties.get(BUCKET_COUNT_PROPERTY);
-        if ((bucketedBy.isEmpty()) && (bucketCount == 0)) {
-            if (!sortedBy.isEmpty()) {
-                throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s may be specified only when %s is specified", SORTED_BY_PROPERTY, BUCKETED_BY_PROPERTY));
-            }
-            return Optional.empty();
+
+        List<String> clusteredBy = getClusteredBy(tableProperties);
+        List<Integer> clusterCount = getClusterCount(tableProperties);
+        List<String> distribution = getDistribution(tableProperties);
+
+        if (doBucketing(bucketedBy)) {
+            return getPropertyForBucketing(bucketedBy, sortedBy, bucketCount);
         }
-        if (bucketCount < 0) {
-            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s must be greater than zero", BUCKET_COUNT_PROPERTY));
+
+        if (doClustering(clusteredBy)) {
+            return getPropertyForClustering(bucketedBy, sortedBy, bucketCount, clusteredBy, clusterCount, distribution);
         }
-        if (bucketCount > 1_000_000) {
-            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s should be no more than 1000000", BUCKET_COUNT_PROPERTY));
+
+        if (!sortedBy.isEmpty()) {
+            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s may be specified only when %s is specified",
+                    SORTED_BY_PROPERTY, BUCKETED_BY_PROPERTY));
         }
-        if (bucketedBy.isEmpty() || bucketCount == 0) {
-            throw new PrestoException(INVALID_TABLE_PROPERTY, format("%s and %s must be specified together", BUCKETED_BY_PROPERTY, BUCKET_COUNT_PROPERTY));
+        return Optional.empty();
+    }
+
+    private static boolean doBucketing(List<String> bucketedBy)
```

Comment:
I think this function and `doClustering` should be inlined. It is a bit too trivial.

Reply:
Ok, it is my personal preference of readability. Not sure if Java compiler can do the automatical inlining; if not, I will change.